File: Chapter05.py
# Data Science with SQL Server Quick Start Guide
# Chapter 05

# Imports
import numpy as np
import pandas as pd
import pyodbc
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
import scipy as sc

# Handling NULLs
con = pyodbc.connect('DSN=AWDW;UID=RUser;PWD=Pa$$w0rd')
query = """SELECT c1, c2, c3
           FROM dbo.NULLTest;"""
NULLTest = pd.read_sql(query, con)
NULLTest

# Checking for NULLs
pd.isnull(NULLTest)

# Omitting
NULLTest.dropna(axis = 'rows')
NULLTest.dropna(axis = 'columns')


# Aggregate functions
NULLTest.c2.mean()
NULLTest.c2.mean(skipna = False)


# Reading the data from SQL Server
con = pyodbc.connect('DSN=AWDW;UID=RUser;PWD=Pa$$w0rd')
query = """SELECT CustomerKey, CommuteDistance,
            TotalChildren, NumberChildrenAtHome, 
            Gender, HouseOwnerFlag,
            NumberCarsOwned, MaritalStatus,
            Age, YearlyIncome, BikeBuyer,
            EnglishEducation AS Education,
            EnglishOccupation AS Occupation
          FROM dbo.vTargetMail"""
TM = pd.read_sql(query, con)

# check the Age
TM["Age"].describe()

# Generating dummies (indicators)
pd.get_dummies(TM.MaritalStatus)
pd.get_dummies(TM.MaritalStatus, prefix = 'TM')

# Create the dummies
TM1 = TM[['MaritalStatus']].join(pd.get_dummies(TM.MaritalStatus, prefix = 'TM'))
TM1.tail(3)

# Show the Age in 20 equal width bins
TM['AgeEWB'] = pd.cut(TM['Age'], 20)
TM['AgeEWB'].value_counts()
pd.crosstab(TM.AgeEWB,
            columns = 'Count') .plot(kind = 'bar',
                                    legend = False,
                                    title = 'AgeEWB20')
plt.show()

# Equal width binning - 5 bins
TM['AgeEWB'] = pd.cut(TM['Age'], 5)
TM['AgeEWB'].value_counts(sort = False)


# Equal height binning
TM['AgeEHB'] = pd.qcut(TM['Age'], 5)
TM['AgeEHB'].value_counts(sort = False)


# Custom binning
custombins = [16, 22, 29, 39, 54, 88]
TM['AgeCUB'] = pd.cut(TM['Age'], custombins)
TM['AgeCUB'].value_counts(sort = False)
pd.crosstab(TM.AgeCUB,
            columns = 'Count') .plot(kind = 'bar',
                                    legend = False,
                                    title = 'AgeCUB')
plt.show()


# Calculating the entropy
# Function that calculates the entropy
def f_entropy(indata):
    indataprob = indata.value_counts() / len(indata) 
    entropy=sc.stats.entropy(indataprob, base = 2) 
    return entropy

# Use the function on variables
f_entropy(TM.NumberCarsOwned), np.log2(5), f_entropy(TM.NumberCarsOwned) / np.log2(5)
f_entropy(TM.BikeBuyer), np.log2(2), f_entropy(TM.BikeBuyer) / np.log2(2)


# rx_data_step filtering variables
from revoscalepy import rx_data_step
TM4 = rx_data_step(input_data=TM.iloc[0:3,], 
                   vars_to_keep = {'CustomerKey', 'Age', 'AgeCUB'})
TM4


###########################
# More code as a bonus:-) #
###########################

# Projections
TM1 = TM[["CustomerKey", "MaritalStatus"]]
TM2 = TM[["CustomerKey", "Gender"]]

# Positional access
TM1.iloc[0:3, 0:2]
TM2.iloc[0:3, 0:2]

# Filter and projection
TM.loc[TM.Age > 85, ["CustomerKey", "Age", "Gender"]]

# Joining data frames
TM3 = pd.merge(TM1, TM2, on = "CustomerKey")
TM3.iloc[0:3, 0:3]

# Sort
TMSortedByAge = TM.sort_values(by = ["Age"], ascending = False)
TMSortedByAge.loc[TM.Age > 84, ["CustomerKey", "Age"]]

# Connecting and reading the data
con = pyodbc.connect('DSN=AWDW;UID=RUser;PWD=Pa$$w0rd')
query = """
SELECT g.EnglishCountryRegionName AS Country,
 c.EnglishEducation AS Education,
 c.YearlyIncome AS Income,
 c.NumberCarsOwned
FROM dbo.DimCustomer AS c
 INNER JOIN dbo.DimGeography AS g
  ON c.GeographyKey = g.GeographyKey;"""
TM = pd.read_sql(query, con)

# Group by
TM.groupby('Country')['Income'].count()
TM.groupby('Country')['Income'].median()
TM.groupby('Country')['Income'].describe()

# Aggregate
TM.groupby('Country').aggregate({'Income': 'std',
                                 'NumberCarsOwned':'mean'})
# Agg is an alias for aggregate
# Multiple functions on multiple columns
TM.groupby('Country').aggregate({'Income': ['max', 'mean'],
                                 'NumberCarsOwned':['sum', 'count']})

# Filtering aggregations
TMAGG = pd.DataFrame(TM.groupby('Country')['Income'].mean())
TMAGG.loc[TMAGG.Income > 50000]

# Plot
ax = TMAGG.plot(kind = 'bar',
           color = ('b', 'y', 'r', 'g', 'm', 'k'),
           fontsize = 14, legend = False, 
           use_index = True, rot = 1)
ax.set_xlabel('Country', fontsize = 16)
ax.set_ylabel('Income Mean', fontsize = 16)
plt.show()

File: Chapter05.R
# Data Science with SQL Server Quick Start Guide
# Chapter 05

# Handling NULLs
library(RODBC)
con <- odbcConnect("AWDW", uid = "RUser", pwd = "Pa$$w0rd")
NULLTest <-
sqlQuery(con,
         "SELECT c1, c2, c3
          FROM dbo.NULLTest;")
close(con)
NULLTest

# Working with NULLs
na.omit(NULLTest)
is.na(NULLTest)

# Aggregate functions
mean(NULLTest$c2)
mean(NULLTest$c2, na.rm=TRUE)


# Data manipulation
con <- odbcConnect("AWDW", uid = "RUser", pwd = "Pa$$w0rd")
TM <-
sqlQuery(con,
         "SELECT CustomerKey, CommuteDistance,
            TotalChildren, NumberChildrenAtHome, 
            Gender, HouseOwnerFlag, MaritalStatus,
            NumberCarsOwned, MaritalStatus,
            Age, YearlyIncome, BikeBuyer,
            EnglishEducation AS Education,
            EnglishOccupation AS Occupation
          FROM dbo.vTargetMail;")
close(con)
# View(TM)


# Get dummies in R
install.packages("dummies")
library(dummies)
# Create the dummies
TM1 <- cbind(TM, dummy(TM$MaritalStatus, sep = "_"))
tail(TM1[c("MaritalStatus", "TM_S", "TM_M")], 3)

# Equal width binning
TM["AgeEWB"] = cut(TM$Age, 5)
table(TM$AgeEWB)

# Equal height binning - a function
EHBinning <- function(data, nofbins) {
    bincases <- rep(length(data) %/% nofbins, nofbins)
    bincases <- bincases + ifelse(1:nofbins <= length(data) %% nofbins, 1, 0)
    bin <- rep(1:nofbins, bincases)
    bin <- bin[rank(data, ties.method = "last")]
    return(factor(bin, levels = 1:nofbins, ordered = TRUE))
}
TM["AgeEHB"] = EHBinning(TM$Age, 5)
table(TM$AgeEHB)

# Histogram from RevoScaleR
library("RevoScaleR")
rxHistogram(formula = ~AgeEHB,
            data = TM)

# Custom binning
TM["AgeCUB"] = cut(TM$Age, c(16, 22, 29, 39, 54, 88))
table(TM$AgeCUB)


# Entropy (install only if needed) from DescTools
install.packages("DescTools")
library("DescTools")
# Entropy
NCO = table(TM$NumberCarsOwned)
print(c(Entropy(NCO), log2(5), Entropy(NCO) / log2(5)))
BBT = table(TM$BikeBuyer)
print(c(Entropy(BBT), log2(2), Entropy(BBT) / log2(2)))



# Connecting and reading the data
con <- odbcConnect("AWDW", uid = "RUser", pwd = "Pa$$w0rd")
TM <- as.data.frame(sqlQuery(con,
  "SELECT c.CustomerKey,
    g.EnglishCountryRegionName AS Country,
    c.EnglishEducation AS Education,
    c.YearlyIncome AS Income,
    c.NumberCarsOwned AS Cars,
    C.MaritalStatus,
    c.NumberChildrenAtHome
   FROM dbo.DimCustomer AS c
    INNER JOIN dbo.DimGeography AS g
     ON c.GeographyKey = g.GeographyKey;"),
  stringsAsFactors = TRUE)
close(con)

# Package dplyr
install.packages("dplyr")
library(dplyr)

# Projection
head(TM)
head(select(TM, Income:MaritalStatus))
head(select(TM, - Income))
head(select(TM, starts_with("C")))

# Filter
# All data frame has 18484 cases
count(TM)
# 2906 cases with more than 2 cars
count(filter(TM, Cars > 2))
# Check
table(TM$Cars)

# Sort
head(arrange(TM, desc(CustomerKey)))

# Pipe operator
TM %>%
select(starts_with("C")) %>%
filter(Cars > 2) %>%
count

# Projection, filter, sort
TM %>%
select(starts_with("C")) %>%
filter(Cars > 2) %>%
arrange(desc(CustomerKey)) %>%
head

# Adding a column
TM %>%
filter(Cars > 0) %>%
mutate(PartnerExists = as.integer(ifelse(MaritalStatus == 'S', 0, 1))) %>%
mutate(HouseHoldNumber = 1 + PartnerExists + NumberChildrenAtHome) %>%
select(CustomerKey, Country, HouseHoldNumber, Cars) %>%
arrange(desc(CustomerKey)) %>%
head

# Aggregating
TM %>%
filter(Cars > 0) %>%
mutate(PartnerExists = as.integer(ifelse(MaritalStatus == 'S', 0, 1))) %>%
mutate(HouseHoldNumber = 1 + PartnerExists + NumberChildrenAtHome) %>%
select(CustomerKey, Country, HouseHoldNumber, Cars) %>%
summarise(avgCars = mean(Cars),
    avgHouseHoldNumber = mean(HouseHoldNumber))

# Grouping and aggregating
TM %>%
filter(Cars > 0) %>%
mutate(PartnerExists = as.integer(ifelse(MaritalStatus == 'S', 0, 1))) %>%
mutate(HouseHoldNumber = 1 + PartnerExists + NumberChildrenAtHome) %>%
select(CustomerKey, Country, HouseHoldNumber, Cars) %>%
group_by(Country) %>%
summarise(avgCars = mean(Cars),
    avgHouseHoldNumber = mean(HouseHoldNumber)) %>%
arrange(desc(Country))

# Storing in a df
TM1 =
TM %>%
filter(Cars > 0) %>%
mutate(PartnerExists = as.integer(ifelse(MaritalStatus == 'S', 0, 1))) %>%
mutate(HouseHoldNumber = 1 + PartnerExists + NumberChildrenAtHome) %>%
select(CustomerKey, Country, HouseHoldNumber, Cars) %>%
group_by(Country) %>%
summarise(avgCars = mean(Cars),
    avgHouseHoldNumber = mean(HouseHoldNumber)) %>%
arrange(desc(Country))
TM1

# Basic scatterplot
plot(TM1$avgCars ~ TM1$avgHouseHoldNumber, cex = 2, lwd = 2)


# Package Car scatterplot
install.packages("car")
library(car)
scatterplot(avgCars ~ avgHouseHoldNumber | Country,
    data = TM1,
    xlab = "HouseHoldNumber Avg", ylab = "Cars Avg",
    main = "Enhanced Scatter Plot",
    cex = 3.5, lwd = 15,
    cex.lab = 1.3,
    xlim = c(2.2, 3.6), ylim = c(1.7, 2.2),
    col = c('red', 'blue', 'green', 'black', 'orange', 'magenta'),
    boxplot = 'xy')


###########################
# More code as a bonus:-) #
###########################

# Equal height binning explained - step by step
# Lower limit for the number of cases in a bin
length(TM$Age) %/% 5
# Create the vector of the bins with number of cases
rep(length(TM$Age) %/% 5, 5)
# How many bins need a case more
length(TM$Age) %% 5
# Array to add cases to the first 4 bins
ifelse(1:5 <= length(TM$Age) %% 5, 1, 0)

# Projections
cols1 <- c("CustomerKey", "MaritalStatus")
TM1 <- TM[cols1]
cols2 <- c("CustomerKey", "Gender")
TM2 <- TM[cols2]
TM1[1:3, 1:2]
TM2[1:3, 1:2]

# Merge datasets
TM3 <- merge(TM1, TM2, by = "CustomerKey")
TM3[1:3, 1:3]

# Binding datasets
TM4 <- cbind(TM1, TM2)
TM4[1:3, 1:4]

# Filtering and row binding data
TM1 <- TM[TM$CustomerKey < 11002, cols1]
TM2 <- TM[TM$CustomerKey > 29481, cols1]
TM5 <- rbind(TM1, TM2)
TM5

# Aggregations in groups
aggregate(TM$Income, by = list(TM$Country), FUN = sum)

# More grouping
aggregate(TM$Income,
    by = list(TM$Country, TM$Education),
    FUN = mean)

# Filtering aggregations
TMAGG <- aggregate(list(Income = TM$Income),
    by = list(Country = TM$Country), FUN = median)
TMAGG[TMAGG$Income > 60000,]

# Plot the income median over countries
barplot(TMAGG$Income,
    legend = TMAGG$Country,
    col = c('blue', 'yellow', 'red', 'green', 'magenta', 'black'))
